kern_interp/quadtree/quadtree.cpp

```diff
@@ -912,62 +1206,234 @@ void QuadTree::populate_half_level_dofs(int level) {
+    std::vector<int> dofs;
+    if (node->compressed) {
+      dofs = node->dof_lists.skel;
+    } else if (node->is_leaf) {
+      dofs = node->dof_lists.original_box;
+    } else {
+      dofs = node->dof_lists.active_box;
+    }
+
+    // Remove hif removed dofs
+    std::vector<int> hif_deactivated_dofs;
+    for (MidLevelNode* recompressor_node : node->recompressor_nodes) {
+      hif_deactivated_dofs.insert(hif_deactivated_dofs.end(),
+                                  recompressor_node->dof_lists.redundant.begin(),
+                                  recompressor_node->dof_lists.redundant.end());
+    }
+    std::sort(hif_deactivated_dofs.begin(), hif_deactivated_dofs.end());
+    std::sort(dofs.begin(), dofs.end());
+    std::vector<int> difference;
+    std::set_difference(dofs.begin(), dofs.end(),
+                        hif_deactivated_dofs.begin(), hif_deactivated_dofs.end(),
+                        std::back_inserter(difference));
+    dofs = difference;
+
+    for (int i = 0; i < dofs.size(); i++) {
+      int idx = dofs[i];
+      int points_vec_index = (idx / solution_dimension) *
+                             domain_dimension;
+      double mindist = node->side_length * 10;
+      std::vector<double> closest_edge_center;
+      // Check edge centers
+
+      // iterate over edges`
+      std::vector<std::vector<double>> edgecenters;
+      for (int i = 0; i < 8; i++) {
+        edgecenters.push_back(node->center);
+      }
+      edgecenters[0][0] -= node->side_length / 2.0;
+      edgecenters[0][1] -= node->side_length / 2.0;
+      edgecenters[1][0] += node->side_length / 2.0;
+      edgecenters[1][1] -= node->side_length / 2.0;
+      edgecenters[2][0] -= node->side_length / 2.0;
+      edgecenters[2][1] += node->side_length / 2.0;
+      edgecenters[3][0] += node->side_length / 2.0;
+      edgecenters[3][1] += node->side_length / 2.0;
+      edgecenters[4][2] -= node->side_length / 2.0;
+      edgecenters[4][1] -= node->side_length / 2.0;
+      edgecenters[5][2] += node->side_length / 2.0;
+      edgecenters[5][1] -= node->side_length / 2.0;
+      edgecenters[6][2] -= node->side_length / 2.0;
+      edgecenters[6][1] += node->side_length / 2.0;
+      edgecenters[7][2] += node->side_length / 2.0;
+      edgecenters[7][1] += node->side_length / 2.0;
+      for (std::vector<double> edgecenter : edgecenters) {
+        double edgedist = 0;
+        for (int dd = 0; dd < domain_dimension; dd++) {
+          edgedist += pow(edgecenter[dd] - boundary_points[points_vec_index + dd], 2);
+        }
+        if (sqrt(edgedist) < mindist) {
+          closest_edge_center = edgecenter;
+          mindist = sqrt(edgedist);
+        }
+      }
+      // Now check if closest edgecenter corresponds to a recomp node
+      for (MidLevelNode* recompressor_third_node : node->recompressor_third_nodes) {
+        if (almost_align(closest_edge_center, recompressor_third_node->center)) {
```

Comment:
isn't this possibly readding dofs to compressed nodes?
check for half level too